[css-contain-intrinsic-size] Introduce last remembered size to Element

cathiechen opened the pull request:
#### 0ef6378c4923a62d3ea1cf172308b9a8ffd2431e
<pre>
[css-contain-intrinsic-size] Introduce last remembered size to Element
<a href="https://bugs.webkit.org/show_bug.cgi?id=248998">https://bugs.webkit.org/show_bug.cgi?id=248998</a>

Reviewed by Cameron McCormack.

Per[1], when contain-intrinsic-size is with the auto-length value, the last remembered size could be its explicit intrinsic inner size.
This patch introduce last remembered size to Element, and the work related to ResizeObserver will be done in the follow-up patches.

[1] <a href="https://w3c.github.io/csswg-drafts/css-sizing-4/#valdef-contain-intrinsic-width-auto-length">https://w3c.github.io/csswg-drafts/css-sizing-4/#valdef-contain-intrinsic-width-auto-length</a>

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::lastRememberedSize const):
(WebCore::Element::setLastRememberedSize):
(WebCore::Element::clearLastRememberedSize):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::lastRememberedSize const): The lastRememberedSize is a RefPtr of ResizeObserverSize, and is stored it in ElementRareData.
(WebCore::ElementRareData::setLastRememberedSize):
(WebCore::ElementRareData::clearLastRememberedSize):
(WebCore::ElementRareData::useTypes const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::explicitIntrinsicInnerWidth const): Handle AutoAndLength and use lastRememberedSize if exist.
(WebCore::RenderBox::explicitIntrinsicInnerHeight const):

Canonical link: <a href="https://commits.webkit.org/258183@main">https://commits.webkit.org/258183@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d5b99d3d4905c5a08de3f345f5bceca866f1e6d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101113 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10263 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34159 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110405 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170655 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105101 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1138 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93515 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108235 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106895 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8474 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91740 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35079 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23137 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78072 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3917 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24654 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3957 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1062 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10063 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44163 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/5621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5717 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cathiechen commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5515,6 +5516,11 @@ std::optional<LayoutUnit> RenderBox::explicitIntrinsicInnerWidth() const
     if (style().containIntrinsicWidthType() == ContainIntrinsicSizeType::None)
         return std::nullopt;
 
+    if (element() && style().containIntrinsicWidthType() == ContainIntrinsicSizeType::AutoAndLength && shouldSkipContent()) {
```
CIS auto needs to call  `shouldSkipContent()`, so it depends on `content-visibility`

cathiechen replied:
See https://w3c.github.io/csswg-drafts/css-sizing-4/#valdef-contain-intrinsic-width-auto-length

heycam commented on Source/WebCore/dom/Element.h:
```diff
@@ -680,6 +681,10 @@ class Element : public ContainerNode {
 
+    ResizeObserverSize* lastRememberedSize() const;
+    void saveLastRememberedSize(Ref<ResizeObserverSize>&&);
+    void removeLastRememberedSize();
```
I think we should use the same naming scheme as on `ElementRareData`, i.e. `setLastRememberedSize` and `clearLastRememberedSize`.

cathiechen replied:
Done

heycam commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4414,6 +4414,23 @@ ResizeObserverData* Element::resizeObserverData()
+void Element::saveLastRememberedSize(Ref<ResizeObserverSize>&& size)
+{
+    auto& rareData = ensureElementRareData();
+    rareData.setLastRememberedSize(WTFMove(size));
```
This can be one line:

```
ensureElementRareData().setLastRememberedSize(WTFMove(size));
```

cathiechen replied:
Done, thanks:)

heycam commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4414,6 +4414,23 @@ ResizeObserverData* Element::resizeObserverData()
+void Element::removeLastRememberedSize()
+{
+    if (lastRememberedSize())
+        elementRareData()->clearLastRememberedSize();
```
Maybe `hasRareData()` instead of `lastRememberedSize()` for the condition?

cathiechen replied:
Done

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5551,6 +5557,11 @@ std::optional<LayoutUnit> RenderBox::explicitIntrinsicInnerHeight() const
+    if (element() && style().containIntrinsicHeightType() == ContainIntrinsicSizeType::AutoAndLength && shouldSkipContent()) {
+        if (auto lastRememberedSize = element()->lastRememberedSize())
+            return isHorizontalWritingMode() ? LayoutUnit(lastRememberedSize->blockSize()) : LayoutUnit(lastRememberedSize->inlineSize());
+    }
```
I think a lambda for this may make things more readable and descriptive.

cathiechen replied:
Hmm, I think the code is only called once in this function

cathiechen committed:
[css-contain-intrinsic-size] Introduce last remembered size to Element
https://bugs.webkit.org/show_bug.cgi?id=248998

Reviewed by Cameron McCormack.

Per[1], when contain-intrinsic-size is with the auto-length value, the last remembered size could be its explicit intrinsic inner size.
This patch introduce last remembered size to Element, and the work related to ResizeObserver will be done in the follow-up patches.

[1] https://w3c.github.io/csswg-drafts/css-sizing-4/#valdef-contain-intrinsic-width-auto-length

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::lastRememberedSize const):
(WebCore::Element::setLastRememberedSize):
(WebCore::Element::clearLastRememberedSize):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::lastRememberedSize const): The lastRememberedSize is a RefPtr of ResizeObserverSize, and is stored it in ElementRareData.
(WebCore::ElementRareData::setLastRememberedSize):
(WebCore::ElementRareData::clearLastRememberedSize):
(WebCore::ElementRareData::useTypes const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::explicitIntrinsicInnerWidth const): Handle AutoAndLength and use lastRememberedSize if exist.
(WebCore::RenderBox::explicitIntrinsicInnerHeight const):

Canonical link: https://commits.webkit.org/258183@main
